### ISSUE TITLE
renamed_files in process metrics does not help identifying renamed files

### ISSUE THREAD

**Issue by hosseinkshvrz** [ishepard/pydriller#190]
It's not clear what the purpose of renamed files is in process metrics classes. The first thing that comes to mind is that it is used to avoid considering renamed files two distinct files but the current implementation is not correct I guess. 

Here's the implementation of the core part of the CommitsCount class in the latest version of PyDriller:

```
def count(self):
    files = {}
    renamed_files = {}  # To keep track of renamed files

    for commit in self.repo_miner.traverse_commits():

        for modified_file in commit.modified_files:

            filepath = renamed_files.get(modified_file.new_path,
                                         modified_file.new_path)
            if modified_file.change_type == ModificationType.RENAME:
                renamed_files[modified_file.old_path] = filepath

            files[filepath] = files.get(filepath, 0) + 1

    return files
```
So first thing is that traversing commits is ordered by time from the oldest to the newest. The process above starts with the oldest commit (of course based on the criteria, not necessarily the oldest commit in the repository) and the files that were modified in this commit. The filepath is either the new filepath or is coming from renamed_files. The thing is renamed_files is updated after this line in a way that the old path is mapped to the new path. since we're going forward in time, there should not be an instance that the old name of a modified file is looked up in the renamed_files dictionary. At least not an expected instance. Am I right?

for example:

        commit 1) file A.
        commit 2) file A.
        commit 3) file A renamed to D.
        commit 4) file D modified.

the example above should show file A (or D) have been changed in 4 commits but it shows file A was changed in 2 commits and file D in 2 commits.

If I'm not mistaken the traversal should be in the reverse order. that way the newer commit is seen, the old name of a file in this commit is added to the dictionary. In the next iteration this old path is the new path of the file so when we get it from the renamed_files, the future name of this file is fetched. 

But even if we reverse the order, this approach partially works (only for files that were renamed once). So I'm not sure if I understood the purpose of renamed_files correctly.

**Comment by stefanodallapalma** [comment]
Hi @hosseinkshvrz 

> It's not clear what the purpose of renamed files is in process metrics classes. The first thing that comes to mind is that it is used to avoid considering renamed files two distinct files but the current implementation is not correct I guess.
> 

Indeed, the purpose of `renamed_files` is to avoid that renamed files are considered as distinct files when analyzing the history **backward**.   

> So first thing is that traversing commits is ordered by time from the oldest to the newest. The process above starts with the oldest commit (of course based on the criteria, not necessarily the oldest commit in the repository) and the files that were modified in this commit. 

Please note that the process above always starts with the oldest commit regardless of the criteria. See the [parent class](https://github.com/ishepard/pydriller/blob/151319fec5ea39e461756dedef848992375055d3/pydriller/metrics/process/process_metric.py#L46). 

>The filepath is either the new filepath or is coming from renamed_files. The thing is renamed_files is updated after this line in a way that the old path is mapped to the new path. 

Exactly. The first time the file is met `filepath=modified_file.new_path`, which is the name of the up-to-date version.
Otherwise, (only) if the files is renamed, `renamed_files` is updated so that the _key_ is the old path and the _value_ is **always** the most recent filepath. 
For example,

```
Commit 8 - create origin/main.py
Commit 9 - rename origin/main.py to main.py
Commit 10 - rename main.py to **pydriller**/main.py
```

The process starts at Commit 10. Here `renamed_files = {}`. Here, _main.py_ is renamed into _**pydriller**/main.py_. However, the latter is not in `renamed_files`, therefore `filepath = pydriller/main.py` and renamed_files is updated in a way that 

```
renamed_files = {
   'main.py': 'pydriller/main.py'
}
```
Then, `files[filepath]` is updated by 1, that is

```
files = {
    'pydriller/main.py': 1
}
```

Next is Commit 9. Here, _origin/main.py_ is renamed into _main.py_. However, the latter (filename at commit 9) is an older version of a previously analyzed file, that is `renamed_files['main.py'] = pydriller/main.py`. Therefore, `filepath = pydriller/main.py` 

Then, `files[filepath]` is updated by 1, that is

```
files = {
    'pydriller/main.py': 2
}
```

Also, because the ModificationType of _main.py_ is RENAME, renamed_files is updated so that 

```
renamed_files = {
   'main.py': 'pydriller/main.py'
   'origin/main.py': 'pydriller/main.py'
}
```

That way, we would expect to encounter the file _'origin/main.py'_ somewhere in the next commits.
When this happens, for example at Commit 8, `filepath = renamed_files['origin/main.py']` will be mapped to 'pydriller/main.py' (note that at Commit 8 `modified_file.new_path = 'origin/main.py'`) and 

```
files = {
    'pydriller/main.py': 3
}
```

And so on so forth until the oldest commit between `to_commit` or `from_commit`. The returned dictionary contains only the files updated in the most recent commit between `to_commit` or `from_commit`.

> Since we're going forward in time, there should not be an instance that the old name of a modified file is looked up in the renamed_files dictionary. At least not an expected instance. Am I right?

Please note that we are going **backward** in time, not forward.


> If I'm not mistaken the traversal should be in the reverse order. that way the newer commit is seen, the old name of a file in this commit is added to the dictionary. In the next iteration this old path is the new path of the file so when we get it from the renamed_files, the future name of this file is fetched.

Ideally, going backward or forward doesn't really matter, as it just affects the implementation but the result is the same. 
The main reason of going backward is that process metrics are calculated only for _modified_files_, not for every files in the repository. Therefore, it makes sense to calculate them for the files that were modified in the most recent commit (based on the passed parameters).
That way, the output dictionary `files<filepath, metric value>` reports values only for those files.

> But even if we reverse the order, this approach partially works (only for files that were renamed once). So I'm not sure if I understood the purpose of renamed_files correctly.

To sum up, the purpose is to map all the previous filepath of a file to the most recent filepath.
Perhaps, for the sake of readability, a good idea would be renaming `renamed_files` to something like `most_recent_filepath_of`. So that  `most_recent_filepath_of['an/older/path] = 'most/recent/path'`.

Hope that helps.

**Comment by hosseinkshvrz** [comment]
Thank you Stefano. You're right. I didn't notice the order in the parent directory is "reverse". That solves the problem.
